## ■ 11장 . 의사결정트리


### 머신러닝의 종류 3가지

1. 지도학습 
    - 분류 : knn, 나이브베이즈, 의사결정트리
    - 회귀 
    
    
2. 비지도학습 
3. 강화학습

## ■ 의사결정트리란?

__Decision Tree 는 의사결정 나무__라는 뜻입니다. 컴퓨터 알고리즘에 즐겨사용하는 __트리(tree)구조__를 사용하고,   
각 분기점(node)에는 분석 대상의 속성(설명변수)들이 위치한다.    
각 분기점 마다 목표값을 가장 잘 분류할 수 있는 속성을 찾아서 배치하고, 해당 속성이 갖는 값을 이용하여 __새로운 가지(branch)__를 만든다.  
각 분기점에서 최저의 속성을 선택할 때는 해당 속성을 기준으로 분류한 값들이 구분되는 정도를 측정한다.  
다른 종류의 값들이 섞여 있는 정도를 나타내느 __엔트로피(entropy)__를 활용하여 __엔트로피가 낮을 수록 분류가 잘된것__이고   
엔트로피가 일정 수준 이하로 낮아질때까지 앞의 과정을 반복합니다.

## ■ 의사결정트리 파이썬 코드 예제

" 유방암 데이터의 악성종양과 양성종양을 예측하는 머신러닝 모델 생성 "

In [4]:
import pandas as pd

# Breast Cancer 데이터셋 가져오기 (출처: UCI ML Repository)

uci_path = 'https://archive.ics.uci.edu/ml/machine-learning-databases/\
breast-cancer-wisconsin/breast-cancer-wisconsin.data'

df = pd.read_csv(uci_path, header=None)



# 열 이름 지정

df.columns = ['id','clump','cell_size','cell_shape', 'adhesion','epithlial',

              'bare_nuclei','chromatin','normal_nucleoli', 'mitoses', 'class'] 

print(df)

          id  clump  cell_size  cell_shape  adhesion  epithlial bare_nuclei  \
0    1000025      5          1           1         1          2           1   
1    1002945      5          4           4         5          7          10   
2    1015425      3          1           1         1          2           2   
3    1016277      6          8           8         1          3           4   
4    1017023      4          1           1         3          2           1   
..       ...    ...        ...         ...       ...        ...         ...   
694   776715      3          1           1         1          3           2   
695   841769      2          1           1         1          2           1   
696   888820      5         10          10         3          7           3   
697   897471      4          8           6         4          3           4   
698   897471      4          8           8         5          4           5   

     chromatin  normal_nucleoli  mitoses  class  
0

## 문제196. 위의 컬럼이 모두다 나오게 설정하시오!

In [7]:
import pandas as pd

# Breast Cancer 데이터셋 가져오기 (출처: UCI ML Repository)

uci_path = 'https://archive.ics.uci.edu/ml/machine-learning-databases/\
breast-cancer-wisconsin/breast-cancer-wisconsin.data'

df = pd.read_csv(uci_path, header=None)



# 열 이름 지정

df.columns = ['id','clump','cell_size','cell_shape', 'adhesion','epithlial',

              'bare_nuclei','chromatin','normal_nucleoli', 'mitoses', 'class'] 

pd.set_option('display.max_columns', 15)
print(df)
print('\n')

          id  clump  cell_size  cell_shape  adhesion  epithlial bare_nuclei  \
0    1000025      5          1           1         1          2           1   
1    1002945      5          4           4         5          7          10   
2    1015425      3          1           1         1          2           2   
3    1016277      6          8           8         1          3           4   
4    1017023      4          1           1         3          2           1   
..       ...    ...        ...         ...       ...        ...         ...   
694   776715      3          1           1         1          3           2   
695   841769      2          1           1         1          2           1   
696   888820      5         10          10         3          7           3   
697   897471      4          8           6         4          3           4   
698   897471      4          8           8         5          4           5   

     chromatin  normal_nucleoli  mitoses  class  
0

### ※ 의사결정 머신러닝 모델을 생성하려면 데이터가 전부 숫자형이어야 하므로 숫자형인지 확인해보아야한다.

## 문제197. 유방암 데이터프레임의 구조를 확인해서 데이터 타입이 전부 숫자형인지 확인하시오!

In [8]:
import pandas as pd

# Breast Cancer 데이터셋 가져오기 (출처: UCI ML Repository)

uci_path = 'https://archive.ics.uci.edu/ml/machine-learning-databases/\
breast-cancer-wisconsin/breast-cancer-wisconsin.data'

df = pd.read_csv(uci_path, header=None)



# 열 이름 지정

df.columns = ['id','clump','cell_size','cell_shape', 'adhesion','epithlial',

              'bare_nuclei','chromatin','normal_nucleoli', 'mitoses', 'class'] 

pd.set_option('display.max_columns', 15)


print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 699 entries, 0 to 698
Data columns (total 11 columns):
id                 699 non-null int64
clump              699 non-null int64
cell_size          699 non-null int64
cell_shape         699 non-null int64
adhesion           699 non-null int64
epithlial          699 non-null int64
bare_nuclei        699 non-null object
chromatin          699 non-null int64
normal_nucleoli    699 non-null int64
mitoses            699 non-null int64
class              699 non-null int64
dtypes: int64(10), object(1)
memory usage: 60.2+ KB
None


### 문제198. 유일하게 숫자형이 아닌 brae_nuclei 컬럼을 숫자형으로 변환하시오!( 힌트 : astype )

In [ ]:
import pandas as pd

# Breast Cancer 데이터셋 가져오기 (출처: UCI ML Repository)

uci_path = 'https://archive.ics.uci.edu/ml/machine-learning-databases/\
breast-cancer-wisconsin/breast-cancer-wisconsin.data'

df = pd.read_csv(uci_path, header=None)



# 열 이름 지정

df.columns = ['id','clump','cell_size','cell_shape', 'adhesion','epithlial',

              'bare_nuclei','chromatin','normal_nucleoli', 'mitoses', 'class'] 

pd.set_option('display.max_columns', 15)


df['bare_nuclei'] = df['bare_nuclei'].astype('int')

print(df.info()) 

# 오류 뜸
# ValueError: invalid literal for int() with base 10: '?'

### 문제199. bare_nuclie 데이터 안에 ?가 있어서 숫자형으로 변환되지 않으므로 ?를 NaN으로 replace하고 누락데이터행을 삭제하시오!

In [17]:
import pandas as pd
import numpy as np
# Breast Cancer 데이터셋 가져오기 (출처: UCI ML Repository)

uci_path = 'https://archive.ics.uci.edu/ml/machine-learning-databases/\
breast-cancer-wisconsin/breast-cancer-wisconsin.data'

df = pd.read_csv(uci_path, header=None)



# 열 이름 지정

df.columns = ['id','clump','cell_size','cell_shape', 'adhesion','epithlial',

              'bare_nuclei','chromatin','normal_nucleoli', 'mitoses', 'class'] 

pd.set_option('display.max_columns', 15)


df['bare_nuclei'].replace('?', np.nan, inplace = True)
df.dropna(subset = ['bare_nuclei'], axis = 0, inplace = True)
df['bare_nuclei'] = df['bare_nuclei'].astype('int')

print(df.info()) 

<class 'pandas.core.frame.DataFrame'>
Int64Index: 683 entries, 0 to 698
Data columns (total 11 columns):
id                 683 non-null int64
clump              683 non-null int64
cell_size          683 non-null int64
cell_shape         683 non-null int64
adhesion           683 non-null int64
epithlial          683 non-null int64
bare_nuclei        683 non-null int32
chromatin          683 non-null int64
normal_nucleoli    683 non-null int64
mitoses            683 non-null int64
class              683 non-null int64
dtypes: int32(1), int64(10)
memory usage: 61.4 KB
None


## 문제200. 독립변수(설명변수)와 종속변수(예측변수)로 데이터를 나누시오~

In [19]:
import pandas as pd
import numpy as np
# Breast Cancer 데이터셋 가져오기 (출처: UCI ML Repository)

uci_path = 'https://archive.ics.uci.edu/ml/machine-learning-databases/\
breast-cancer-wisconsin/breast-cancer-wisconsin.data'

df = pd.read_csv(uci_path, header=None)



# 열 이름 지정

df.columns = ['id','clump','cell_size','cell_shape', 'adhesion','epithlial',

              'bare_nuclei','chromatin','normal_nucleoli', 'mitoses', 'class'] 

pd.set_option('display.max_columns', 15)


df['bare_nuclei'].replace('?', np.nan, inplace = True)
df.dropna(subset = ['bare_nuclei'], axis = 0, inplace = True)
df['bare_nuclei'] = df['bare_nuclei'].astype('int')

print(df.columns)

X = df[['clump', 'cell_size', 'cell_shape', 'adhesion', 'epithlial',
       'bare_nuclei', 'chromatin', 'normal_nucleoli', 'mitoses']] # 설명변수

y = df['class'] # 예측변수


print(y)

Index(['id', 'clump', 'cell_size', 'cell_shape', 'adhesion', 'epithlial',
       'bare_nuclei', 'chromatin', 'normal_nucleoli', 'mitoses', 'class'],
      dtype='object')
0      2
1      2
2      2
3      2
4      2
      ..
694    2
695    2
696    4
697    4
698    4
Name: class, Length: 683, dtype: int64


## 문제201. 설명변수 데이터를 정규화 하시오!

In [ ]:
import pandas as pd
import numpy as np

# Breast Cancer 데이터셋 가져오기 (출처: UCI ML Repository)

uci_path = 'https://archive.ics.uci.edu/ml/machine-learning-databases/\
breast-cancer-wisconsin/breast-cancer-wisconsin.data'

df = pd.read_csv(uci_path, header=None)



# 열 이름 지정

df.columns = ['id','clump','cell_size','cell_shape', 'adhesion','epithlial',

              'bare_nuclei','chromatin','normal_nucleoli', 'mitoses', 'class'] 

pd.set_option('display.max_columns', 15)


df['bare_nuclei'].replace('?', np.nan, inplace = True)
df.dropna(subset = ['bare_nuclei'], axis = 0, inplace = True)
df['bare_nuclei'] = df['bare_nuclei'].astype('int')


X = df[['clump', 'cell_size', 'cell_shape', 'adhesion', 'epithlial',
       'bare_nuclei', 'chromatin', 'normal_nucleoli', 'mitoses']] # 설명변수

y = df['class'] # 예측변수

from sklearn import preprocessing

X = preprocessing.StandardScaler().fit(X).transform(X)

## 문제202. 기존 데이터를 훈련 데이터와 테스트 데이터로 7:3 비율로 나누시오.

In [ ]:
# 추가 코드
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 15)

In [ ]:
import pandas as pd
import numpy as np

# Breast Cancer 데이터셋 가져오기 (출처: UCI ML Repository)

uci_path = 'https://archive.ics.uci.edu/ml/machine-learning-databases/\
breast-cancer-wisconsin/breast-cancer-wisconsin.data'

df = pd.read_csv(uci_path, header=None)



# 열 이름 지정

df.columns = ['id','clump','cell_size','cell_shape', 'adhesion','epithlial',

              'bare_nuclei','chromatin','normal_nucleoli', 'mitoses', 'class'] 

pd.set_option('display.max_columns', 15)


df['bare_nuclei'].replace('?', np.nan, inplace = True)
df.dropna(subset = ['bare_nuclei'], axis = 0, inplace = True)
df['bare_nuclei'] = df['bare_nuclei'].astype('int')


X = df[['clump', 'cell_size', 'cell_shape', 'adhesion', 'epithlial',
       'bare_nuclei', 'chromatin', 'normal_nucleoli', 'mitoses']] # 설명변수

y = df['class'] # 예측변수

from sklearn import preprocessing

X = preprocessing.StandardScaler().fit(X).transform(X)

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 10)

print('train data 의 갯수:', X_train.shape) # (499, 9)
print('test data 의 갯수:', X_test.shape) # (215, 9)
print('\n')

## 문제203. 전처리한 데이터를 의사결정트리 모델에 넣어 예측하시오!

In [32]:
# 추가 코드
from sklearn import tree
tree_model = tree.DecisionTreeClassifier(criterion = 'entropy', max_depth = 5)
tree_model.fit(X_train, y_train)

y_hat = tree_model.predict(X_test)

print(y_hat[0:10])
print(y_test.values[0:10])

[4 4 4 4 4 4 2 2 4 4]
[4 4 4 4 4 4 2 2 4 4]


#### ※ 설명 :   
분류 정도를 평가하는 기준으로 __'entrophy'값__을 사용한다.   
트리레벨을 5로 지정하여 5단계까지 가지를 확장할 수 있게 하겠다.  
레벨이 많아질수록 모형학습에 사용하는 훈련데이터에 대한 예측은 정확해진다.  
하지만 모형이 훈련 데이터에 대해서만 지나치게 최적화 되어서 실제 데이터의 예측 능력은 떨어지는 문제가 발생한다.(Overfitting)  
그래서 __적절한 레벨값__을 찾는것이 중요하다.

In [31]:
# Full Code.
import pandas as pd
import numpy as np

# Breast Cancer 데이터셋 가져오기 (출처: UCI ML Repository)

uci_path = 'https://archive.ics.uci.edu/ml/machine-learning-databases/\
breast-cancer-wisconsin/breast-cancer-wisconsin.data'

df = pd.read_csv(uci_path, header=None)



# 열 이름 지정

df.columns = ['id','clump','cell_size','cell_shape', 'adhesion','epithlial',

              'bare_nuclei','chromatin','normal_nucleoli', 'mitoses', 'class'] 

pd.set_option('display.max_columns', 15)


df['bare_nuclei'].replace('?', np.nan, inplace = True)
df.dropna(subset = ['bare_nuclei'], axis = 0, inplace = True)
df['bare_nuclei'] = df['bare_nuclei'].astype('int')


X = df[['clump', 'cell_size', 'cell_shape', 'adhesion', 'epithlial',
       'bare_nuclei', 'chromatin', 'normal_nucleoli', 'mitoses']] # 설명변수

y = df['class'] # 예측변수

from sklearn import preprocessing

X = preprocessing.StandardScaler().fit(X).transform(X)

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 10)

print('train data 의 갯수:', X_train.shape) # (499, 9)
print('test data 의 갯수:', X_test.shape) # (215, 9)
print('\n')

from sklearn import tree
tree_model = tree.DecisionTreeClassifier(criterion = 'entropy', max_depth = 5)
tree_model.fit(X_train, y_train)

y_hat = tree_model.predict(X_test)

print(y_hat[0:10])
print(y_test.values[0:10])

train data 의 갯수: (478, 9)
test data 의 갯수: (205, 9)


[4 4 4 4 4 4 2 2 4 4]
[4 4 4 4 4 4 2 2 4 4]


## 문제204. 혼동행렬을 그려서 의사결정트리의 성능을 확인하시오!

In [34]:
# 추가 코드
from sklearn import metrics
tree_matrix = metrics.confusion_matrix(y_test,y_hat)
print(tree_matrix)
tn,fp,fn,tp = metrics.confusion_matrix(y_test,y_hat).ravel()
print(tn,fp,fn,tp)

[[127   4]
 [  2  72]]


In [ ]:
# Full Code.
import pandas as pd
import numpy as np

# Breast Cancer 데이터셋 가져오기 (출처: UCI ML Repository)

uci_path = 'https://archive.ics.uci.edu/ml/machine-learning-databases/\
breast-cancer-wisconsin/breast-cancer-wisconsin.data'

df = pd.read_csv(uci_path, header=None)



# 열 이름 지정

df.columns = ['id','clump','cell_size','cell_shape', 'adhesion','epithlial',

              'bare_nuclei','chromatin','normal_nucleoli', 'mitoses', 'class'] 

pd.set_option('display.max_columns', 15)


df['bare_nuclei'].replace('?', np.nan, inplace = True)
df.dropna(subset = ['bare_nuclei'], axis = 0, inplace = True)
df['bare_nuclei'] = df['bare_nuclei'].astype('int')


X = df[['clump', 'cell_size', 'cell_shape', 'adhesion', 'epithlial',
       'bare_nuclei', 'chromatin', 'normal_nucleoli', 'mitoses']] # 설명변수

y = df['class'] # 예측변수

from sklearn import preprocessing

X = preprocessing.StandardScaler().fit(X).transform(X)

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 10)

print('train data 의 갯수:', X_train.shape) # (499, 9)
print('test data 의 갯수:', X_test.shape) # (215, 9)
print('\n')

from sklearn import tree
tree_model = tree.DecisionTreeClassifier(criterion = 'entropy', max_depth = 5)
tree_model.fit(X_train, y_train)

y_hat = tree_model.predict(X_test)

print(y_hat[0:10])
print(y_test.values[0:10])

# 추가 코드
from sklearn import metrics

tree_matrix = metrics.confusion_matrix(y_test,y_hat)
print(tree_matrix)
tn,fp,fn,tp = metrics.confusion_matrix(y_test,y_hat).ravel()
print(tn,fp,fn,tp)

In [35]:
# f1 - score 추가

tree_report = metrics.classification_report(y_test,y_hat)
print(tree_report)

              precision    recall  f1-score   support

           2       0.98      0.97      0.98       131
           4       0.95      0.97      0.96        74

    accuracy                           0.97       205
   macro avg       0.97      0.97      0.97       205
weighted avg       0.97      0.97      0.97       205



In [ ]:
# Full Code.
import pandas as pd
import numpy as np

# Breast Cancer 데이터셋 가져오기 (출처: UCI ML Repository)

uci_path = 'https://archive.ics.uci.edu/ml/machine-learning-databases/\
breast-cancer-wisconsin/breast-cancer-wisconsin.data'

df = pd.read_csv(uci_path, header=None)



# 열 이름 지정

df.columns = ['id','clump','cell_size','cell_shape', 'adhesion','epithlial',

              'bare_nuclei','chromatin','normal_nucleoli', 'mitoses', 'class'] 

pd.set_option('display.max_columns', 15)


df['bare_nuclei'].replace('?', np.nan, inplace = True)
df.dropna(subset = ['bare_nuclei'], axis = 0, inplace = True)
df['bare_nuclei'] = df['bare_nuclei'].astype('int')


X = df[['clump', 'cell_size', 'cell_shape', 'adhesion', 'epithlial',
       'bare_nuclei', 'chromatin', 'normal_nucleoli', 'mitoses']] # 설명변수

y = df['class'] # 예측변수

from sklearn import preprocessing

X = preprocessing.StandardScaler().fit(X).transform(X)

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 10)

print('train data 의 갯수:', X_train.shape) # (499, 9)
print('test data 의 갯수:', X_test.shape) # (215, 9)
print('\n')

from sklearn import tree
tree_model = tree.DecisionTreeClassifier(criterion = 'entropy', max_depth = 5)
tree_model.fit(X_train, y_train)

y_hat = tree_model.predict(X_test)

print(y_hat[0:10])
print(y_test.values[0:10])

# 추가 코드
from sklearn import metrics

tree_matrix = metrics.confusion_matrix(y_test,y_hat)
print(tree_matrix)
tn,fp,fn,tp = metrics.confusion_matrix(y_test,y_hat).ravel()
print(tn,fp,fn,tp)

# f1 - score 추가

tree_report = metrics.classification_report(y_test,y_hat)
print(tree_report)

### 코드 정리

1. 전처리 작업  
?를 결측치로 치환하고 결측치 제거 (699개를 행에서 16개의 행을 삭제하고 683개의 행으로 훈련)  
문자형 --> 숫자형

2. 전처리 작업
정규화 작업 수행  

3. 모델 : 의사결정트리 모델(엔진 : entropy, random_state = 33, max_depth = 3)

4. accuracy = 0.97

In [ ]:
# 악성종양 코드 변수를 바꾸어서 정확도 0.98까지 올려본 코드
# Full Code.
import pandas as pd
import numpy as np

# Breast Cancer 데이터셋 가져오기 (출처: UCI ML Repository)

uci_path = 'https://archive.ics.uci.edu/ml/machine-learning-databases/\
breast-cancer-wisconsin/breast-cancer-wisconsin.data'

df = pd.read_csv(uci_path, header=None)



# 열 이름 지정

df.columns = ['id','clump','cell_size','cell_shape', 'adhesion','epithlial',

              'bare_nuclei','chromatin','normal_nucleoli', 'mitoses', 'class'] 

pd.set_option('display.max_columns', 15)


df['bare_nuclei'].replace('?', np.nan, inplace = True)
df.dropna(subset = ['bare_nuclei'], axis = 0, inplace = True)
df['bare_nuclei'] = df['bare_nuclei'].astype('int')


X = df[['clump', 'cell_size', 'cell_shape', 'adhesion', 'epithlial',
       'bare_nuclei', 'chromatin', 'normal_nucleoli', 'mitoses']] # 설명변수

y = df['class'] # 예측변수

from sklearn import preprocessing

X = preprocessing.StandardScaler().fit(X).transform(X)

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 33)

print('train data 의 갯수:', X_train.shape) # (499, 9)
print('test data 의 갯수:', X_test.shape) # (215, 9)
print('\n')

from sklearn import tree
tree_model = tree.DecisionTreeClassifier(criterion = 'entropy', max_depth = 3)
tree_model.fit(X_train, y_train)

y_hat = tree_model.predict(X_test)

print(y_hat[0:10])
print(y_test.values[0:10])

# 추가 코드
from sklearn import metrics

tree_matrix = metrics.confusion_matrix(y_test,y_hat)
print(tree_matrix)
tn,fp,fn,tp = metrics.confusion_matrix(y_test,y_hat).ravel()
print(tn,fp,fn,tp)

# f1 - score 추가

tree_report = metrics.classification_report(y_test,y_hat)
print(tree_report)

from sklearn.metrics import accuracy_score
accuracy = accuracy_score( y_test, y_hat).round(5)
print(accuracy)

## 문제205. skin.csv 데이터를 가지고 쿠폰 반응이 있는 고객과 없는 고객을 예측하는 의사결정트리 모델을 생성하시오!

In [40]:
# 악성종양 코드 변수를 바꾸어서 정확도 0.98까지 올려본 코드
# Full Code.
import pandas as pd
import numpy as np

# Breast Cancer 데이터셋 가져오기 (출처: UCI ML Repository)
df = pd.read_csv("d:\\data\\skin.csv", encoding = 'cp949') # 한글이 있기때문에 cp949 로 encoding을 해주어야한다.
print(df)
print('\n')

print(df.info())
print('\n')

    cust_no  gender  age  job marry  car cupon_react
0         1    male   30   NO   YES   NO          NO
1         2  female   20  YES   YES  YES          NO
2         3  female   20  YES   YES   NO          NO
3         4  female   40   NO    NO   NO          NO
4         5  female   30   NO   YES   NO          NO
5         6  female   30   NO    NO  YES          NO
6         7  female   20   NO   YES   NO          NO
7         8  female   20   NO   YES  YES         YES
8         9  female   30  YES   YES   NO         YES
9        10    male   40  YES    NO  YES         YES
10       11    male   20   NO    NO   NO          NO
11       12    male   30   NO   YES  YES          NO
12       13    male   20  YES    NO   NO          NO
13       14  female   30  YES   YES   NO         YES
14       15    male   30  YES   YES  YES         YES
15       16  female   30  YES    NO   NO          NO
16       17  female   30   NO   YES  YES         YES
17       18    male   20  YES   YES   NO      

### 데이터 전처리 작업 : 문자 ---> 숫자(더미변수 생성)
예 : df['gender'] = pd.get_dummies(df['gender'])

In [1]:
# 악성종양 코드 변수를 바꾸어서 정확도 0.98까지 올려본 코드
# Full Code.
import pandas as pd
import numpy as np

# Breast Cancer 데이터셋 가져오기 (출처: UCI ML Repository)
df = pd.read_csv("d:\\data\\skin.csv", encoding = 'cp949') # 한글이 있기때문에 cp949 로 encoding을 해주어야한다.
print(df)
print('\n')

print(df.info())
print('\n')

print(df.columns)

df['gender'] = pd.get_dummies(df['gender'])
df['job'] = pd.get_dummies(df['job'])
df['marry'] = pd.get_dummies(df['marry'])
df['car'] = pd.get_dummies(df['car'])
df['cupon_react'] = pd.get_dummies(df['cupon_react'])

X = df[['gender', 'age', 'job', 'marry', 'car']] # 설명변수

y = df['cupon_react'] # 예측변수

from sklearn import preprocessing

X = preprocessing.StandardScaler().fit(X).transform(X)

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 34)

print('train data 의 갯수:', X_train.shape) # (499, 9)
print('test data 의 갯수:', X_test.shape) # (215, 9)
print('\n')

from sklearn import tree
tree_model = tree.DecisionTreeClassifier(criterion = 'entropy', max_depth = 3)
tree_model.fit(X_train, y_train)

y_hat = tree_model.predict(X_test)

print(y_hat[0:10])
print(y_test.values[0:10])

# 추가 코드
from sklearn import metrics

tree_matrix = metrics.confusion_matrix(y_test,y_hat)
print(tree_matrix)
tn,fp,fn,tp = metrics.confusion_matrix(y_test,y_hat).ravel()
print(tn,fp,fn,tp)

# f1 - score 추가

tree_report = metrics.classification_report(y_test,y_hat)
print(tree_report)

from sklearn.metrics import accuracy_score
accuracy = accuracy_score( y_test, y_hat).round(5)
print(accuracy)

    cust_no  gender  age  job marry  car cupon_react
0         1    male   30   NO   YES   NO          NO
1         2  female   20  YES   YES  YES          NO
2         3  female   20  YES   YES   NO          NO
3         4  female   40   NO    NO   NO          NO
4         5  female   30   NO   YES   NO          NO
5         6  female   30   NO    NO  YES          NO
6         7  female   20   NO   YES   NO          NO
7         8  female   20   NO   YES  YES         YES
8         9  female   30  YES   YES   NO         YES
9        10    male   40  YES    NO  YES         YES
10       11    male   20   NO    NO   NO          NO
11       12    male   30   NO   YES  YES          NO
12       13    male   20  YES    NO   NO          NO
13       14  female   30  YES   YES   NO         YES
14       15    male   30  YES   YES  YES         YES
15       16  female   30  YES    NO   NO          NO
16       17  female   30   NO   YES  YES         YES
17       18    male   20  YES   YES   NO      

## 문제206. (점심시간 문제) seaborn의 타이타닉 데이터의 생존자/사망자를 예측하는 의사결정트리 모델을 생성하고 혼동행렬을 출력하여 검사받으시오!

In [2]:
# Full code.
import seaborn as sns
import pandas as pd

# 1단계. csv --> 데이터 프레임으로 변환
df = sns.load_dataset('titanic')

# 컬럼이 모두다 출력될 수 있도록 출력할 열의 개수 한도를 늘리기
pd.set_option('display.max_columns', 15)
print(df.head(), '\n')

mask1 = (df['age'] < 10) | (df['sex'] == 'female')
df['child_women'] = mask1.astype(int)

# 2.1 결측치를 확인하고 제거하거나 치환한다.
# 타이타닉 데이터 프레임의 자료형을 확인한다.

print(df.info()) # 자료형 확인하는 방법
print('\n')
# 2.2 결측치(NaN)을 확인한다.
print(df.isnull().sum(axis=0), '\n')

# 설명 : deck 결측치가 많아서 컬럼을 삭제해야함
# embark 와 embark_town이 같은 데이터(중복데이터)여서 컬럼을 하나 삭제해야함.
# embark 컬럼은 삭제한다.

# 2.3 deck 컬럼과 embark_town 컬럼을 삭제한다.
rdf = df.drop(['deck', 'embark_town'], axis =1)
print(rdf.columns.values, '\n')

# 2.4 age(나이) 열에 나이가 없는 모든 행을 삭제한다.
# age 열에 891개의 행중에 177개가 NaN값

rdf = rdf.dropna(subset = ['age'], how='any', axis = 0)
print(len(rdf), '\n')

# 설명
# 데이터가 한개라도 없으면 drop 해라 (how = 'any')
# 모든 데이터가 없으면 drop 해라 (how = 'all')

# 2.5 embark 열의 NaN값을 승선도시중 가장 많이 출현한 값으로 치환하기.

most_freq = rdf['embarked'].value_counts(dropna=True).idxmax()
rdf['embarked'].fillna(most_freq, inplace = True)
print(rdf.isnull().sum(axis=0), '\n')

# 3. 범주형 데이터를 숫자형으로 변환하기

# 3.1 feature selection (분석에 필요한 속성을 선택)
ndf = rdf[['survived','pclass','sex','age','sibsp','parch','embarked', 'child_women']]
print(ndf.head(), '\n')
# 선택된 컬럼중 2개(sex, embarked)가 범주형(문자형)이다.

# 3.2 범주형 데이터를 숫자로 변환(원핫 인코딩)
gender = pd.get_dummies(ndf['sex'])
ndf = pd.concat([ndf,gender], axis = 1)

onehot_embarked = pd.get_dummies(ndf['embarked'], prefix = 'town') # 접두사를 정함.
ndf = pd.concat([ndf, onehot_embarked], axis = 1)

ndf.drop(['sex', 'embarked'], axis = 1 , inplace = True)
print(ndf.head())
print('\n')

# 4단계. 정규화
# survived  pclass   age  sibsp  parch  female  male  town_C  town_Q  town_S
# survived (종속변수 y)
# pclass   age  sibsp  parch  female  male  town_C  town_Q  town_S (독립변수 X)

# 4.1 독립변수와 종속변수(라벨)를 지정한다.
X = ndf[['pclass','age','sibsp','parch','female','male','town_C','town_Q','town_S', 'child_women']] # 독립변수
y = ndf['survived'] # 종속변수

# 4.2 독립변수들을 정규화(normalization) 한다.
from sklearn import preprocessing

X = preprocessing.StandardScaler().fit(X).transform(X)
print(X)
print('\n')

# 5단계 . 데이터셋을 훈련데이터와 테스트 데이터로 나눈다.

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 17)

# 설명 : test_size = 0.3 에 의해서 7:3 비율로 훈련과 테스트를 나누고 
# random_state = 10 에 의해서 나중에 split할 때도 항상 일정하게 split 할 수 있게 한다.

print('train data 의 갯수:', X_train.shape) # (499,9)
print('test data 의 갯수:', X_test.shape) # (215, 9)
print('\n')

# 6단계. 머신러닝 모델을 생성한다.(나이브 베이즈 분류모형을 사용)

from sklearn import tree
tree_model = tree.DecisionTreeClassifier(criterion = 'entropy', max_depth = 3)
tree_model.fit(X_train, y_train)

# 7단계. 테스트 데이터로 예측하기

y_hat = tree_model.predict(X_test)

print(y_hat[0:10])
print(y_test.values[0:10])


# 8단계. 모형의 예측 능력을 평가한다. (confusion matrix 만들기!)
from sklearn import metrics
tree_matrix = metrics.confusion_matrix(y_test,y_hat)
print(tree_matrix)
print('\n')

"""
             예측
           사망 생존
실제 사망 [[109  16]
     생존  [ 25  65]]
     
"""

# 정확도 : (109 + 65) / (109 + 16 + 25 + 65)
# 정확도 100%는 현실적으로 나오기 힘드니 (예측: 생존 , 실제: 사망 데이터인 25를 0으로 만드는 것을 궁극적으로 추구한다.)

# confusion matrix에 대한 변수 지정 (tn,fp,fn,tp)
tn, fp, fn, tp = metrics.confusion_matrix( y_test, y_hat ).ravel()
print(tn,fp,fn,tp)
print('\n')

f1_report = metrics.classification_report(y_test, y_hat)
print(f1_report,'\n')

# 9단계 정확도 확인
from sklearn.metrics import accuracy_score
accuracy = accuracy_score( y_test, y_hat)
print(accuracy)

   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True   

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 16 columns):
survived       891 non-null int64
pclass         891 non-null int64
sex            891 non-null object
age      

## ■ 12장. 랜덤포레스트(Random forest)

### " 의사결정트리 + 앙상블 기법 " 모델

## ■ 앙상블 기법이란 ?

다양한 전문가 팀을 만드는 것과 유사한 원리를 활용하는 메타학습 방법을 __앙상블__이라고 합니다.  
앙상블 방법은 약한 학습자 여러개를 결합하면 강한 학습자가 만들어진다는 아이디어를 기반으로 합니다.  
__앙상블 모형__은 여러개의 분류 모형을 같이 사용하여 한꺼번에 평가를 하는 모형을 말합니다.


### ※ 예제  
정확도가 60%밖에 되지 않는 분류기 모형들이 즐비한데 이 모형들은 최소한 몇개를 써야 정확도를 90%를 능가하게 만들 수 있을까?  

Rstudio에서 r코드로 확인!!!

In [ ]:
ret_err <- function(n,err) {
  sum <- 0 
  for(i in floor(n/2):n) { 
    sum <- sum + choose(n,i) * err^i * (1-err)^(n-i)
  }
  sum
}



for(j in 1:60) {
  err <- ret_err(j , 0.4)
  cat(j,'--->',1-err,'\n') 
  if(1-err >= 0.9) break
}

## ■ 앙상블로 구현한 머신러닝의 종류 2가지

### 1. 배깅(bagging) :   

테스트 데이터 샘플링(boostrap)을 통해 여러개의 테스트 데이터를 만들고,  
각 테스트 데이터를 이용하여 여러개의 week learner을 만들어 종합하는 앙상블의 한 종류

### 2. 부스팅(boosting) : 

boost "격려하다, 향상시키다"  
부스팅은 배깅을 약간 더 개선시킨 알고리즘인데 샘플링하는 과정에서 복원추출을 할 때 동일한 확률로 하는게 아니라   
추출할때마다 확률을 서로 다르게 개선시키는 방법을 쓴다.  

처음에는 모두 동일한 확률로 복원추출하지만  
다음번 추출과정에서는 오분류된 데이터를 추출확률이 더 높도록 조정하고 올바르게 분류된 데이터는 추출 확률을 낮게 조정하여 복원 추출을 한다.

## ■ 랜덤 포레스트(random forests)란?

### decision tree 와 bagging을 결합한 알고리즘이다!!! (decision tree + bagging)


### 랜덤 포레스트는 의사결정트리가 여러개 이다. 이 여러개가 결합해서 강한 학습자가 되는 구조이다.

In [ ]:
# 예제

from sklearn.ensemble import RandomForestClassifier

tree_model = RandomForestClassifier(n_estimators = 100, oob_score = True, random_state = 1234)

### ※ 설명

#### 1. n_estimaors = 100 (생성할 tree(약한학습자)를 100개로 하겠다.)

#### 2. oob_score = True ( out of bag 기능을 사용하겠다.)

out of back --> 100개의 tree가 훈련 데이터를 사용할 때 63%를 사용하고 나머지 37%의 oob 샘플로 평가하겠다는 뜻이다!  
앙상블 평가는 oob 평가를 평균으로 하여 얻습니다.  

oob_score = True로 지정하면 훈련이 끝난후에 자동으로 oob평가를 수행합니다. 평가 결과를 보고싶으면 아래의 명령어를 쓴다.  
print(tree_model.oob_score_)  

### 예제1. 의사결정트리만 이용하였을 때
### 예제2. 랜덤포레스트를 사용하였을 때

In [63]:
# 예제1. 의사결정트리만 이용하였을 때
# 악성종양 코드 변수를 바꾸어서 정확도 0.98까지 올려본 코드
# Full Code.
import pandas as pd
import numpy as np

# Breast Cancer 데이터셋 가져오기 (출처: UCI ML Repository)

uci_path = 'https://archive.ics.uci.edu/ml/machine-learning-databases/\
breast-cancer-wisconsin/breast-cancer-wisconsin.data'

df = pd.read_csv(uci_path, header=None)



# 열 이름 지정

df.columns = ['id','clump','cell_size','cell_shape', 'adhesion','epithlial',

              'bare_nuclei','chromatin','normal_nucleoli', 'mitoses', 'class'] 

pd.set_option('display.max_columns', 15)


df['bare_nuclei'].replace('?', np.nan, inplace = True)
df.dropna(subset = ['bare_nuclei'], axis = 0, inplace = True)
df['bare_nuclei'] = df['bare_nuclei'].astype('int')


X = df[['clump', 'cell_size', 'cell_shape', 'adhesion', 'epithlial',
       'bare_nuclei', 'chromatin', 'normal_nucleoli', 'mitoses']] # 설명변수

y = df['class'] # 예측변수

from sklearn import preprocessing

X = preprocessing.StandardScaler().fit(X).transform(X)

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 33)

print('train data 의 갯수:', X_train.shape) # (499, 9)
print('test data 의 갯수:', X_test.shape) # (215, 9)
print('\n')

from sklearn import tree
tree_model = tree.DecisionTreeClassifier(criterion = 'entropy', max_depth = 3)
tree_model.fit(X_train, y_train)

y_hat = tree_model.predict(X_test)

print(y_hat[0:10])
print(y_test.values[0:10])

# 추가 코드
from sklearn import metrics

tree_matrix = metrics.confusion_matrix(y_test,y_hat)
print(tree_matrix)
tn,fp,fn,tp = metrics.confusion_matrix(y_test,y_hat).ravel()
print(tn,fp,fn,tp)

# f1 - score 추가

tree_report = metrics.classification_report(y_test,y_hat)
print(tree_report)

from sklearn.metrics import accuracy_score
accuracy = accuracy_score( y_test, y_hat).round(5)
print(accuracy)

train data 의 갯수: (478, 9)
test data 의 갯수: (205, 9)


[2 2 4 4 2 2 2 4 2 2]
[2 2 4 4 2 2 2 4 2 2]
[[119   3]
 [  0  83]]
119 3 0 83
              precision    recall  f1-score   support

           2       1.00      0.98      0.99       122
           4       0.97      1.00      0.98        83

    accuracy                           0.99       205
   macro avg       0.98      0.99      0.98       205
weighted avg       0.99      0.99      0.99       205

0.98537


In [ ]:
# 예제2. 랜덤포레스트를 사용하였을 때

from sklearn.ensemble import RandomForestClassifier

tree_model = RandomForestClassifier(n_estimators = 100, oob_score = True, random_state = 1234)

In [78]:
# 예제1. 의사결정트리만 이용하였을 때
# 악성종양 코드 변수를 바꾸어서 정확도 0.98까지 올려본 코드
# Full Code.
import pandas as pd
import numpy as np

# Breast Cancer 데이터셋 가져오기 (출처: UCI ML Repository)

uci_path = 'https://archive.ics.uci.edu/ml/machine-learning-databases/\
breast-cancer-wisconsin/breast-cancer-wisconsin.data'

df = pd.read_csv(uci_path, header=None)



# 열 이름 지정

df.columns = ['id','clump','cell_size','cell_shape', 'adhesion','epithlial',

              'bare_nuclei','chromatin','normal_nucleoli', 'mitoses', 'class'] 

pd.set_option('display.max_columns', 15)


df['bare_nuclei'].replace('?', np.nan, inplace = True)
df.dropna(subset = ['bare_nuclei'], axis = 0, inplace = True)
df['bare_nuclei'] = df['bare_nuclei'].astype('int')


X = df[['clump', 'cell_size', 'cell_shape', 'adhesion', 'epithlial',
       'bare_nuclei', 'chromatin', 'normal_nucleoli', 'mitoses']] # 설명변수

y = df['class'] # 예측변수

from sklearn import preprocessing

X = preprocessing.StandardScaler().fit(X).transform(X)

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 33)

print('train data 의 갯수:', X_train.shape) # (499, 9)
print('test data 의 갯수:', X_test.shape) # (215, 9)
print('\n')

from sklearn.ensemble import RandomForestClassifier

tree_model = RandomForestClassifier(n_estimators = 100, oob_score = True, random_state = 12)



tree_model.fit(X_train, y_train)

# oob 평가
print(tree_model.oob_score_)
print('\n')

y_hat = tree_model.predict(X_test)

print(y_hat[0:10])
print(y_test.values[0:10])
print('\n')

# 추가 코드
from sklearn import metrics

tree_matrix = metrics.confusion_matrix(y_test,y_hat)
print(tree_matrix)
print('\n')

tn,fp,fn,tp = metrics.confusion_matrix(y_test,y_hat).ravel()
print(tn,fp,fn,tp)
print('\n')

# f1 - score 추가

tree_report = metrics.classification_report(y_test,y_hat)
print(tree_report)

from sklearn.metrics import accuracy_score
accuracy = accuracy_score( y_test, y_hat).round(5)
print('정확도는 = ', accuracy)

train data 의 갯수: (478, 9)
test data 의 갯수: (205, 9)


0.9644351464435147


[2 2 4 4 2 2 2 4 2 2]
[2 2 4 4 2 2 2 4 2 2]


[[121   1]
 [  1  82]]


121 1 1 82


              precision    recall  f1-score   support

           2       0.99      0.99      0.99       122
           4       0.99      0.99      0.99        83

    accuracy                           0.99       205
   macro avg       0.99      0.99      0.99       205
weighted avg       0.99      0.99      0.99       205

정확도는 =  0.99024


## 문제207. seaborn의 타이타닉 데이터의 생존자/사망자 예측 모델을 랜덤포레스트로 구현하시오!

In [88]:
# Full code.
import seaborn as sns
import pandas as pd

# 1단계. csv --> 데이터 프레임으로 변환
df = sns.load_dataset('titanic')

# 컬럼이 모두다 출력될 수 있도록 출력할 열의 개수 한도를 늘리기
pd.set_option('display.max_columns', 15)
print(df.head(), '\n')

mask1 = (df['age'] < 10) | (df['sex'] == 'female')
df['child_women'] = mask1.astype(int)

# 2.1 결측치를 확인하고 제거하거나 치환한다.
# 타이타닉 데이터 프레임의 자료형을 확인한다.

print(df.info()) # 자료형 확인하는 방법
print('\n')
# 2.2 결측치(NaN)을 확인한다.
print(df.isnull().sum(axis=0), '\n')

# 설명 : deck 결측치가 많아서 컬럼을 삭제해야함
# embark 와 embark_town이 같은 데이터(중복데이터)여서 컬럼을 하나 삭제해야함.
# embark 컬럼은 삭제한다.

# 2.3 deck 컬럼과 embark_town 컬럼을 삭제한다.
rdf = df.drop(['deck', 'embark_town'], axis =1)
print(rdf.columns.values, '\n')

# 2.4 age(나이) 열에 나이가 없는 모든 행을 삭제한다.
# age 열에 891개의 행중에 177개가 NaN값

rdf = rdf.dropna(subset = ['age'], how='any', axis = 0)
print(len(rdf), '\n')

# 설명
# 데이터가 한개라도 없으면 drop 해라 (how = 'any')
# 모든 데이터가 없으면 drop 해라 (how = 'all')

# 2.5 embark 열의 NaN값을 승선도시중 가장 많이 출현한 값으로 치환하기.

most_freq = rdf['embarked'].value_counts(dropna=True).idxmax()
rdf['embarked'].fillna(most_freq, inplace = True)
print(rdf.isnull().sum(axis=0), '\n')

# 3. 범주형 데이터를 숫자형으로 변환하기

# 3.1 feature selection (분석에 필요한 속성을 선택)
ndf = rdf[['survived','pclass','sex','age','sibsp','parch','embarked', 'child_women']]
print(ndf.head(), '\n')
# 선택된 컬럼중 2개(sex, embarked)가 범주형(문자형)이다.

# 3.2 범주형 데이터를 숫자로 변환(원핫 인코딩)
gender = pd.get_dummies(ndf['sex'])
ndf = pd.concat([ndf,gender], axis = 1)

onehot_embarked = pd.get_dummies(ndf['embarked'], prefix = 'town') # 접두사를 정함.
ndf = pd.concat([ndf, onehot_embarked], axis = 1)

ndf.drop(['sex', 'embarked'], axis = 1 , inplace = True)
print(ndf.head())
print('\n')

# 4단계. 정규화
# survived  pclass   age  sibsp  parch  female  male  town_C  town_Q  town_S
# survived (종속변수 y)
# pclass   age  sibsp  parch  female  male  town_C  town_Q  town_S (독립변수 X)

# 4.1 독립변수와 종속변수(라벨)를 지정한다.
X = ndf[['pclass','age','sibsp','parch','female','male','town_C','town_Q','town_S', 'child_women']] # 독립변수
y = ndf['survived'] # 종속변수

# 4.2 독립변수들을 정규화(normalization) 한다.
from sklearn import preprocessing

X = preprocessing.StandardScaler().fit(X).transform(X)


# 5단계 . 데이터셋을 훈련데이터와 테스트 데이터로 나눈다.

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 14)

# 설명 : test_size = 0.3 에 의해서 7:3 비율로 훈련과 테스트를 나누고 
# random_state = 10 에 의해서 나중에 split할 때도 항상 일정하게 split 할 수 있게 한다.


# 6단계. 머신러닝 모델을 생성한다.(랜덤 포레스트를 사용)

from sklearn.ensemble import RandomForestClassifier

tree_model = RandomForestClassifier(n_estimators = 29, oob_score = True, random_state = 7)


tree_model.fit(X_train, y_train)

# oob 평가
"""
print(tree_model.oob_score_)
print('\n')
"""

# 7단계. 테스트 데이터로 예측하기

y_hat = tree_model.predict(X_test)


# 8단계. 모형의 예측 능력을 평가한다. (confusion matrix 만들기!)
from sklearn import metrics
tree_matrix = metrics.confusion_matrix(y_test,y_hat)

"""
             예측
           사망 생존
실제 사망 [[109  16]
     생존  [ 25  65]]
     
"""

# 정확도 : (109 + 65) / (109 + 16 + 25 + 65)
# 정확도 100%는 현실적으로 나오기 힘드니 (예측: 생존 , 실제: 사망 데이터인 25를 0으로 만드는 것을 궁극적으로 추구한다.)

# confusion matrix에 대한 변수 지정 (tn,fp,fn,tp)
tn, fp, fn, tp = metrics.confusion_matrix( y_test, y_hat ).ravel()


f1_report = metrics.classification_report(y_test, y_hat)
print(f1_report,'\n')

# 9단계 정확도 확인
from sklearn.metrics import accuracy_score
accuracy = accuracy_score( y_test, y_hat)
print(accuracy)

   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True   

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 16 columns):
survived       891 non-null int64
pclass         891 non-null int64
sex            891 non-null object
age      

## ■ 랜덤 포레스트 모델의 성능을 올리기 위하여 데이터 전처리 하는 방법

1. 결측치가 많은 컬럼을 삭제 -- cabin
2. 나이의 결측치를 치환 - 최빈값으로 치환 : 0.80861 상위 6% 진입
3. 이상치를 제거 - 운임(fare)의 이상치를 제거 : 0.81339 상위 5% 진입

## 문제208. seaborn의 타이타닉 데이터에서 결측치가 가장 많은 컬럼인 deck을 삭제하시오!

케글의 타이타닉 데이터의 cabin 컬럼은 객실번호인데  
시본의 타이타닉 데이터의 deck과 같은 컬럼이다.  

In [6]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("titanic")
df2 = pd.read_csv("d:\\data\\tat\\train.csv")

print(df.deck)
print(df2.Cabin)
print(df.deck.isnull().sum())


# 이미 위의 코드에 deck 컬럼을 삭제하는 코드가 구현되어 있다.( 참고만 할 것 !)
rdf = df.drop(['deck','embark_town'], axis =1)

0      NaN
1        C
2      NaN
3        C
4      NaN
      ... 
886    NaN
887      B
888    NaN
889      C
890    NaN
Name: deck, Length: 891, dtype: category
Categories (7, object): [A, B, C, D, E, F, G]
0       NaN
1       C85
2       NaN
3      C123
4       NaN
       ... 
886     NaN
887     B42
888     NaN
889    C148
890     NaN
Name: Cabin, Length: 891, dtype: object
688


## 문제209. 훈련 데이터의 나이의 결측치를 최빈값으로 치환하시오!

In [ ]:
# ※설명 : seaborn 타이타닉에서는 24살이 최빈값.
age_freq = rdf['age'].value_counts(dropna=True).idxmax()
print(age_freq)
print('\n')

rdf['age'].fillna(age_freq, inplace = True)

# ※설명 : seaborn 타이타닉에서는 24살이 최빈값.

In [7]:
# 나이의 결측치를 바꾼 최빈값.
# Full code.

import seaborn as sns
import pandas as pd

# 1단계. csv --> 데이터 프레임으로 변환
df = sns.load_dataset('titanic')

# 컬럼이 모두다 출력될 수 있도록 출력할 열의 개수 한도를 늘리기
pd.set_option('display.max_columns', 15)
# print(df.head(), '\n')

mask1 = (df['age'] < 10) | (df['sex'] == 'female')
df['child_women'] = mask1.astype(int)

# 2.1 결측치를 확인하고 제거하거나 치환한다.
# 타이타닉 데이터 프레임의 자료형을 확인한다.

# print(df.info()) # 자료형 확인하는 방법
# print('\n')

# 2.2 결측치(NaN)을 확인한다.
print(df.isnull().sum(axis=0), '\n')

# 설명 : deck 결측치가 많아서 컬럼을 삭제해야함
# embark 와 embark_town이 같은 데이터(중복데이터)여서 컬럼을 하나 삭제해야함.
# embark 컬럼은 삭제한다.

# 2.3 deck 컬럼과 embark_town 컬럼을 삭제한다.
rdf = df.drop(['deck', 'embark_town'], axis =1)
# print(rdf.columns.values, '\n')

# 2.4 age(나이) 열에 나이가 없는 모든 행을 삭제한다.
# age 열에 891개의 행중에 177개가 NaN값

age_freq = rdf['age'].value_counts(dropna=True).idxmax()
print(age_freq)
print('\n')

print(df.isnull().sum())
print('\n')

rdf['age'].fillna(age_freq, inplace = True)

print(len(rdf), '\n')
print(rdf.isnull().sum())
print('\n')


# 2.5 embark 열의 NaN값을 승선도시중 가장 많이 출현한 값으로 치환하기.

most_freq = rdf['embarked'].value_counts(dropna=True).idxmax()
rdf['embarked'].fillna(most_freq, inplace = True)
print(rdf.isnull().sum(axis=0), '\n')

# 3. 범주형 데이터를 숫자형으로 변환하기

# 3.1 feature selection (분석에 필요한 속성을 선택)
ndf = rdf[['survived','pclass','sex','age','sibsp','parch','embarked', 'child_women']]
# print(ndf.head(), '\n')
# 선택된 컬럼중 2개(sex, embarked)가 범주형(문자형)이다.

# 3.2 범주형 데이터를 숫자로 변환(원핫 인코딩)
gender = pd.get_dummies(ndf['sex'])
ndf = pd.concat([ndf,gender], axis = 1)

onehot_embarked = pd.get_dummies(ndf['embarked'], prefix = 'town') # 접두사를 정함.
ndf = pd.concat([ndf, onehot_embarked], axis = 1)

ndf.drop(['sex', 'embarked'], axis = 1 , inplace = True)
# print(ndf.head())
# print('\n')

# 4단계. 정규화
# survived  pclass   age  sibsp  parch  female  male  town_C  town_Q  town_S
# survived (종속변수 y)
# pclass   age  sibsp  parch  female  male  town_C  town_Q  town_S (독립변수 X)

# 4.1 독립변수와 종속변수(라벨)를 지정한다.
X = ndf[['pclass','age','sibsp','parch','female','male','town_C','town_Q','town_S', 'child_women']] # 독립변수
y = ndf['survived'] # 종속변수

# 4.2 독립변수들을 정규화(normalization) 한다.
from sklearn import preprocessing

X = preprocessing.StandardScaler().fit(X).transform(X)


# 5단계 . 데이터셋을 훈련데이터와 테스트 데이터로 나눈다.

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 19)

# 설명 : test_size = 0.3 에 의해서 7:3 비율로 훈련과 테스트를 나누고 
# random_state = 10 에 의해서 나중에 split할 때도 항상 일정하게 split 할 수 있게 한다.


# 6단계. 머신러닝 모델을 생성한다.(랜덤 포레스트를 사용)

from sklearn.ensemble import RandomForestClassifier

tree_model = RandomForestClassifier(n_estimators = 100, oob_score = True, random_state = 9)


tree_model.fit(X_train, y_train)

# oob 평가
"""
print(tree_model.oob_score_)
print('\n')
"""

# 7단계. 테스트 데이터로 예측하기

y_hat = tree_model.predict(X_test)


# 8단계. 모형의 예측 능력을 평가한다. (confusion matrix 만들기!)
from sklearn import metrics
tree_matrix = metrics.confusion_matrix(y_test,y_hat)

# confusion matrix에 대한 변수 지정 (tn,fp,fn,tp)
tn, fp, fn, tp = metrics.confusion_matrix( y_test, y_hat ).ravel()


f1_report = metrics.classification_report(y_test, y_hat)
print(f1_report,'\n')

# 9단계 정확도 확인
from sklearn.metrics import accuracy_score
accuracy = accuracy_score( y_test, y_hat)
print(accuracy)

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
child_women      0
dtype: int64 

24.0


survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
child_women      0
dtype: int64


891 

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       2
class          0
who            0
adult_male     0
alive          0
alone          0
child_women    0
dtype: int64


survived       0
pclass         0
sex            0
age            0
sibsp          0
parch      

## ■ 운임(fare)의 이상치를 제거하는 방법

## 문제210. 운임의 이상치를 확인하시오!

In [94]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset('titanic')
print(df.fare.mean())

32.2042079685746


## 문제211. 운임 데이터의 표준편차를 확인하시오!

In [97]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset('titanic')
print(df.fare.std())

49.693428597180905


## 문제212. 운임 데이터의 최대값, 최소값, 평균값을 한번에 출력하시오!

In [98]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset('titanic')
print(df.fare.describe())

count    891.000000
mean      32.204208
std       49.693429
min        0.000000
25%        7.910400
50%       14.454200
75%       31.000000
max      512.329200
Name: fare, dtype: float64


## 문제213. 아래의 SQL을 파이썬으로 구현하시오!

In [101]:
"""
select fare
    from (select fare,
                round(5*stddev(fare) over (order by null) local_std
            from titanic)
         )
        
    where fare > local_std;
"""

import seaborn as sns
df = sns.load_dataset('titanic')
local_std = df.fare.std() * 5
result = df['fare'][df['fare']> local_std]

print(result)
print('\n')
print(len(result))

27     263.0000
88     263.0000
258    512.3292
311    262.3750
341    263.0000
438    263.0000
679    512.3292
737    512.3292
742    262.3750
Name: fare, dtype: float64


9


## 문제214. 위의 9명을 이상치라고 했을 때 이상치를 포함했을때의 운임 평균과 이상치를 제거했을때의 운임평균이 각각 어떻게 되는지 확인하시오!

In [102]:
# 이상치 제거전

import seaborn as sns
df1 = sns.load_dataset('titanic')
print(df1.fare.mean())

32.2042079685746


In [107]:
# 이상치 제거후

import seaborn as sns
df= sns.load_dataset('titanic')
local_std = df.fare.std() * 5 # 이상치의 조건 --> 표준편차의 5배가 되는 값

result = df['fare'][df['fare'] < local_std] # 이상치를 제거하는 코드 !!
print(result)
print(result.mean())

0       7.2500
1      71.2833
2       7.9250
3      53.1000
4       8.0500
        ...   
886    13.0000
887    30.0000
888    23.4500
889    30.0000
890     7.7500
Name: fare, Length: 882, dtype: float64
29.002507596371835


## 문제215. 운임의 이상치(9건)을 제거하고 랜덤 포레스트 모델을 생성해서 정확도를 확인하시오!

In [108]:
# 이상치제거
# 나이값 최빈값
# 랜덤 포레스트
# Full Code.ver1 (반복문 x)

# 나이의 결측치를 바꾼 최빈값.
# Full code.

import seaborn as sns
import pandas as pd

# 1단계. csv --> 데이터 프레임으로 변환
df = sns.load_dataset('titanic')

# 컬럼이 모두다 출력될 수 있도록 출력할 열의 개수 한도를 늘리기
pd.set_option('display.max_columns', 15)
print(df.head(), '\n')

mask1 = (df['age'] < 10) | (df['sex'] == 'female')
df['child_women'] = mask1.astype(int)

# 2.1 결측치를 확인하고 제거하거나 치환한다.
# 타이타닉 데이터 프레임의 자료형을 확인한다.

print(df.info()) # 자료형 확인하는 방법
print('\n')

# 2.2 결측치(NaN)을 확인한다.
print(df.isnull().sum(axis=0), '\n')


# 설명 : deck 결측치가 많아서 컬럼을 삭제해야함
# embark 와 embark_town이 같은 데이터(중복데이터)여서 컬럼을 하나 삭제해야함.
# embark 컬럼은 삭제한다.

# 2.3 deck 컬럼과 embark_town 컬럼을 삭제한다.
rdf = df.drop(['deck', 'embark_town'], axis =1)
# print(rdf.columns.values, '\n')

# 2.4 age(나이) 열에 나이가 없는 모든 행을 삭제한다.
# age 열에 891개의 행중에 177개가 NaN값

age_freq = rdf['age'].value_counts(dropna=True).idxmax()
print(age_freq)
print('\n')

print(df.isnull().sum(axis=0))
print('\n')

rdf['age'].fillna(age_freq, inplace = True)

print(len(rdf), '\n')
print(rdf.isnull().sum(axis=0))
print('\n')

# 운임값(fare)의 이상치를 제거
local_std = rdf['fare'].std() * 5
rdf = rdf[:][rdf['fare'] < local_std]

print(rdf.isnull().sum(axis=0), '\n')

# 운임값(fare)의 결측치를 최빈값으로 채움.
fare_freq = rdf['fare'].value_counts(dropna=True).idxmax()
rdf['fare'].fillna(fare_freq, inplace = True)


# 2.5 embark 열의 NaN값을 승선도시중 가장 많이 출현한 값으로 치환하기.

most_freq = rdf['embarked'].value_counts(dropna=True).idxmax()
rdf['embarked'].fillna(most_freq, inplace = True)
print(rdf.isnull().sum(axis=0), '\n')

# 3. 범주형 데이터를 숫자형으로 변환하기

# 3.1 feature selection (분석에 필요한 속성을 선택)
ndf = rdf[['survived','pclass','sex','age','sibsp','parch','embarked', 'child_women', 'fare']]
# print(ndf.head(), '\n')
# 선택된 컬럼중 2개(sex, embarked)가 범주형(문자형)이다.

# 3.2 범주형 데이터를 숫자로 변환(원핫 인코딩)
gender = pd.get_dummies(ndf['sex'])
ndf = pd.concat([ndf,gender], axis = 1)

onehot_embarked = pd.get_dummies(ndf['embarked'], prefix = 'town') # 접두사를 정함.
ndf = pd.concat([ndf, onehot_embarked], axis = 1)

ndf.drop(['sex', 'embarked'], axis = 1 , inplace = True)
# print(ndf.head())
# print('\n')

# 4단계. 정규화
# survived  pclass   age  sibsp  parch  female  male  town_C  town_Q  town_S
# survived (종속변수 y)
# pclass   age  sibsp  parch  female  male  town_C  town_Q  town_S (독립변수 X)

# 4.1 독립변수와 종속변수(라벨)를 지정한다.
X = ndf[['pclass','age','sibsp','parch','female','male','town_C','town_Q','town_S', 'child_women', 'fare']] # 독립변수
y = ndf['survived'] # 종속변수

# 4.2 독립변수들을 정규화(normalization) 한다.
from sklearn import preprocessing

X = preprocessing.StandardScaler().fit(X).transform(X)


# 5단계 . 데이터셋을 훈련데이터와 테스트 데이터로 나눈다.

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 19)

# 설명 : test_size = 0.3 에 의해서 7:3 비율로 훈련과 테스트를 나누고 
# random_state = 10 에 의해서 나중에 split할 때도 항상 일정하게 split 할 수 있게 한다.


# 6단계. 머신러닝 모델을 생성한다.(랜덤 포레스트를 사용)

from sklearn.ensemble import RandomForestClassifier

tree_model = RandomForestClassifier(n_estimators = 100, oob_score = True, random_state = 9)


tree_model.fit(X_train, y_train)

# oob 평가
"""
print(tree_model.oob_score_)
print('\n')
"""

# 7단계. 테스트 데이터로 예측하기

y_hat = tree_model.predict(X_test)


# 8단계. 모형의 예측 능력을 평가한다. (confusion matrix 만들기!)
from sklearn import metrics
tree_matrix = metrics.confusion_matrix(y_test,y_hat)

# confusion matrix에 대한 변수 지정 (tn,fp,fn,tp)
tn, fp, fn, tp = metrics.confusion_matrix( y_test, y_hat ).ravel()


f1_report = metrics.classification_report(y_test, y_hat)
print(f1_report,'\n')

# 9단계 정확도 확인
from sklearn.metrics import accuracy_score
accuracy = accuracy_score( y_test, y_hat)
print(accuracy)

   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True   

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 16 columns):
survived       891 non-null int64
pclass         891 non-null int64
sex            891 non-null object
age      

## 위의코드에서 반복문까지 추가해서 작성해보시오!

In [ ]:
# Full code.
import seaborn as sns
import pandas as pd

# 1단계. csv --> 데이터 프레임으로 변환
df = sns.load_dataset('titanic')

# 컬럼이 모두다 출력될 수 있도록 출력할 열의 개수 한도를 늘리기
pd.set_option('display.max_columns', 15)
print(df.head(), '\n')

mask1 = (df['age'] < 10) | (df['sex'] == 'female')
df['child_women'] = mask1.astype(int)

# 2.1 결측치를 확인하고 제거하거나 치환한다.
# 타이타닉 데이터 프레임의 자료형을 확인한다.

print(df.info()) # 자료형 확인하는 방법
print('\n')
# 2.2 결측치(NaN)을 확인한다.
print(df.isnull().sum(axis=0), '\n')

# 설명 : deck 결측치가 많아서 컬럼을 삭제해야함
# embark 와 embark_town이 같은 데이터(중복데이터)여서 컬럼을 하나 삭제해야함.
# embark 컬럼은 삭제한다.

# 2.3 deck 컬럼과 embark_town 컬럼을 삭제한다.
rdf = df.drop(['deck', 'embark_town'], axis =1)
print(rdf.columns.values, '\n')

# 2.4 age(나이) 열에 나이가 없는 모든 행을 삭제한다.
# age 열에 891개의 행중에 177개가 NaN값

age_freq = rdf['age'].value_counts(dropna=True).idxmax()
print(age_freq)
print('\n')

rdf['age'].fillna(age_freq, inplace = True)


# 운임값(fare)의 이상치를 제거
local_std = rdf['fare'].std() * 5
rdf = rdf[:][rdf['fare'] < local_std]

print(rdf.isnull().sum(axis=0), '\n')

# 운임값(fare)의 결측치를 최빈값으로 채움.
fare_freq = rdf['fare'].value_counts(dropna=True).idxmax()
rdf['fare'].fillna(fare_freq, inplace = True)

# 설명
# 데이터가 한개라도 없으면 drop 해라 (how = 'any')
# 모든 데이터가 없으면 drop 해라 (how = 'all')

# 2.5 embark 열의 NaN값을 승선도시중 가장 많이 출현한 값으로 치환하기.

most_freq = rdf['embarked'].value_counts(dropna=True).idxmax()
rdf['embarked'].fillna(most_freq, inplace = True)
print(rdf.isnull().sum(axis=0), '\n')

# 3. 범주형 데이터를 숫자형으로 변환하기

# 3.1 feature selection (분석에 필요한 속성을 선택)
ndf = rdf[['survived','pclass','sex','age','sibsp','parch','embarked', 'child_women', 'fare']]
print(ndf.head(), '\n')
# 선택된 컬럼중 2개(sex, embarked)가 범주형(문자형)이다.

# 3.2 범주형 데이터를 숫자로 변환(원핫 인코딩)
gender = pd.get_dummies(ndf['sex'])
ndf = pd.concat([ndf,gender], axis = 1)

onehot_embarked = pd.get_dummies(ndf['embarked'], prefix = 'town') # 접두사를 정함.
ndf = pd.concat([ndf, onehot_embarked], axis = 1)

ndf.drop(['sex', 'embarked'], axis = 1 , inplace = True)
print(ndf.head())
print('\n')

# 4단계. 정규화
# survived  pclass   age  sibsp  parch  female  male  town_C  town_Q  town_S
# survived (종속변수 y)
# pclass   age  sibsp  parch  female  male  town_C  town_Q  town_S (독립변수 X)

# 4.1 독립변수와 종속변수(라벨)를 지정한다.
X = ndf[['pclass','age','sibsp','parch','female','male','town_C','town_Q','town_S', 'child_women', 'fare']] # 독립변수
y = ndf['survived'] # 종속변수

# 4.2 독립변수들을 정규화(normalization) 한다.
from sklearn import preprocessing

X = preprocessing.StandardScaler().fit(X).transform(X)

# 5단계 . 데이터셋을 훈련데이터와 테스트 데이터로 나눈다.
result = []
for i in range(20,30,1):
    from sklearn.model_selection import train_test_split
    
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = i)
    
    # 설명 : test_size = 0.3 에 의해서 7:3 비율로 훈련과 테스트를 나누고 
    # random_state = 10 에 의해서 나중에 split할 때도 항상 일정하게 split 할 수 있게 한다.
    
    
    # 6단계. 머신러닝 모델을 생성한다.(랜덤 포레스트를 사용)
    for j in range(10,50,2):
        from sklearn.ensemble import RandomForestClassifier
        
        tree_model = RandomForestClassifier(n_estimators = 100, oob_score = True, random_state = j)
        
        
        tree_model.fit(X_train, y_train)
    
        # oob 평가
        print(tree_model.oob_score_)
        print('\n')
        
        # 7단계. 테스트 데이터로 예측하기
        
        y_hat = tree_model.predict(X_test)
        
        
        # 8단계. 모형의 예측 능력을 평가한다. (confusion matrix 만들기!)
        from sklearn import metrics
        tree_matrix = metrics.confusion_matrix(y_test,y_hat)
        
        """
                     예측
                   사망 생존
        실제 사망 [[109  16]
             생존  [ 25  65]]
             
        """
        
        # 정확도 : (109 + 65) / (109 + 16 + 25 + 65)
        # 정확도 100%는 현실적으로 나오기 힘드니 (예측: 생존 , 실제: 사망 데이터인 25를 0으로 만드는 것을 궁극적으로 추구한다.)
        
        # confusion matrix에 대한 변수 지정 (tn,fp,fn,tp)
        tn, fp, fn, tp = metrics.confusion_matrix( y_test, y_hat ).ravel()
        
        
        f1_report = metrics.classification_report(y_test, y_hat)
        
        # 9단계 정확도 확인
        from sklearn.metrics import accuracy_score
        accuracy = accuracy_score( y_test, y_hat)
        
        print(i,j, end = '\n\n')
        print(accuracy)
        result.append(accuracy)

print('\n')
print('최대 accuracy는 ', max(result))
print('최대값의 위치는 = ', result.index(max(result)))

## 선생님이 짜신 seaborn 타이타닉 모범코드

In [109]:

from sklearn import metrics
import numpy as np
# 1단계 csv ---> 데이터 프레임으로 변환

import pandas as pd
import seaborn as sns

df = sns.load_dataset('titanic')
#df2 = pd.read_csv("d://data//train.csv")
#print(df.head())
#print(df2.Cabin)



# 컬럼이 모두다 출력될 수 있도록 출력할 열의 개수 한도를 늘리기

pd.set_option('display.max_columns',15)



# 2단계 결측치 확인하고 제거하거나 치환한다.

# 2.1 타이타닉 데이터 프레임의 자료형을 확인한다.

mask4 = (df.age<10) | (df.sex=='female') 

df['child_women']=mask4.astype(int)

# 2.2 결측치(NaN) 을 확인한다.


# 2.3 deck 컬럼과 embark_town 컬럼을 삭제한다.

# 설명 : deck 결측치가 많아서 컬럼을 삭제해야함.

#        embark 와 embark_town 이 같은 데이터여서 embark 컬럼을 삭제해야함


rdf = df.drop(['deck','embark_town'], axis =1)

most_freq = rdf['age'].value_counts(dropna=True).idxmax()

print(most_freq)

print ( df.isnull().sum( axis=0)  )  

rdf['age'].fillna(most_freq, inplace = True)

print ( rdf.isnull().sum( axis=0)  )  


# 2.4 age(나이) 열에 나이가 없는 모든행을 삭제한다.

# 데이터가 한개라도 없으면 drop 해라 (how = 'any')

# 모든 데이터가 없으면 drop 해라 (how = 'all')

#rdf = rdf.dropna( subset=['age'], how='any', axis=0)


# 2.5 embark 열의 NaN 값을 승선도시중 가장 많이 출현한 값으로 치환하기


most_freq = rdf['embarked'].value_counts().idxmax()

rdf['embarked'].fillna(most_freq, inplace = True)


# 3단계 범주형 데이터를 숫자형으로 변환하기


# 3.1 feature selection (분석에 필요한 속성을 선택)

ndf = rdf[['survived','fare', 'pclass','sex','age','sibsp','parch','embarked','child_women']]



# 선택된 컬럼중 2개(sex, embarked) 가 범주형이다.
#3.2 범주형 데이터를 숫자로 변환하기(원핫 인코딩)


gender = pd.get_dummies(ndf['sex'])

ndf = pd.concat([ndf,gender], axis= 1)
onehot_embarked = pd.get_dummies(ndf['embarked'])
ndf = pd.concat([ndf,onehot_embarked],axis=1)
ndf.drop(['sex','embarked'], axis=1, inplace = True)


#most_freq = ndf['fare'].value_counts().idxmax()
#ndf['fare'].fillna(most_freq, inplace = True)
local_std = ndf.fare.std() * 5
print(local_std)
ndf = ndf[:][ndf['fare'] < local_std]

print(ndf)


# 4단계 정규화



# 4.1 독립변수와 종속변수(라벨) 을 지정한다.

# survived  pclass   age  sibsp  parch  female  male  C  Q  S

#   라벨                       데이터

# 종속변수                     독립변수

x = ndf[ ['fare', 'pclass', 'age' ,'sibsp', 'parch' ,'female' ,'male', 'C' ,'Q' ,'S',
          'child_women'] ]

y = ndf['survived'] # 종속변수



# 4.2 독립변수들을 정규화 한다.

from sklearn import preprocessing

X = preprocessing.StandardScaler().fit(x).transform(x)


# 5단계 훈련 데이터를 훈련 데이터 / 테스트 데이터로 나눈다

from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test = train_test_split(X,y,test_size = 0.3,
                                                 random_state = 14)


# sklearn 라이브러리에서 나이브베이즈 분류 모형 가져오기


from  sklearn.ensemble   import  RandomForestClassifier 

tree_model = RandomForestClassifier( n_estimators=100, oob_score=True, 
                                     random_state=9)  


tree_model.fit( X_train, y_train )

print ( tree_model.oob_score_)


# 7단계 테스트 데이터로 예측을 한다.

y_hat = tree_model.predict( X_test )



# 8단계 모형의 예측능력을 평가한다.

from sklearn import metrics

randomforest_matrix = metrics.confusion_matrix( y_test, y_hat )

print( randomforest_matrix )



tn, fp, fn, tp = metrics.confusion_matrix( y_test, y_hat ).ravel()



f1_report = metrics.classification_report( y_test, y_hat )

print( f1_report )

#print(np.array([[tp,fp],[fn,tn]]))


from sklearn.metrics import accuracy_score

accuracy = accuracy_score( y_test, y_hat)

print(accuracy)

24.0
survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
child_women      0
dtype: int64
survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       2
class          0
who            0
adult_male     0
alive          0
alone          0
child_women    0
dtype: int64
248.46714298590453
     survived     fare  pclass   age  sibsp  parch  child_women  female  male  \
0           0   7.2500       3  22.0      1      0            0       0     1   
1           1  71.2833       1  38.0      1      0            1       1     0   
2           1   7.9250       3  26.0      0      0            1       1     0   
3           1  53.1000       1  35.0      1      0            1       1     0   
4  

### 문제216.(오늘의 마지막 문제) 케글의 타이타닉 데이터를 랜덤포레스트 모델을 써서 테스트 데이터의 예측값을 구해서 자신의 순위를 올리시오~

1. 나이의 결측치를 최빈값으로 치환
    - 최빈값으로 치환할 때 훈련데이터는 훈련데이터의 최빈값으로 변경 (31살)
    - 최빈값으로 치환할 때 테스트 데이터는 테스트 데이터의 최빈값으로 변경(21살)
    
    
2. 운임의 이상치를 훈련데이터에서만 제거하시오 ~ ( 5* 표준편차)